crates/ide-diagnostics/src/handlers/trait_impl_reduntant_assoc_item.rs

```diff
@@ -0,0 +1,88 @@
+use hir::{db::ExpandDatabase, Const, Function, HasSource, TypeAlias};
+use syntax::ast;
+
+use crate::{Diagnostic, DiagnosticCode, DiagnosticsContext};
+
+// Diagnostic: trait-impl-reduntant-assoc_item
+//
+// Diagnoses reduntant trait items in a trait impl.
+pub(crate) fn trait_impl_reduntant_assoc_item(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::TraitImplReduntantAssocItems,
+) -> Diagnostic {
+    let name = d.assoc_item.0.clone();
+    let assoc_item = d.assoc_item.1;
+    let db = ctx.sema.db;
+
+    let root = ctx.sema.db.parse_or_expand(d.file_id);
+    let range = root.text_range();
+    let impl_ = d.impl_.to_node(&root);
+    let trait_name = if let Some(trait_name) = trait_name(ctx, impl_) {
+        trait_name
+    } else {
+        String::from("{unkown}")
+    };
```

Comment:
We have the `hir::Trait` when constructing the diagnostic already, so we might as well just pass that into the `TraitImplRedundantAssocItems` struct so we can fetch its name through the hir here directly instead of going through the AST